src/common/download.cpp

```diff
@@ -40,27 +41,86 @@
 
 namespace tools
 {
-  static bool download_thread(const std::string &path, const std::string &url)
+  struct download_thread_control
   {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool stop;
+    bool stopped;
+    bool success;
+    std::shared_ptr<boost::thread> thread;
+    boost::mutex mutex;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), stop(false), stopped(false), success(false) {}
+    void lock() { mutex.lock(); }
+    void unlock() { mutex.unlock(); }
+  };
+
+  static void download_thread(download_thread_control *control)
+  {
+    static std::atomic<unsigned int> thread_id(0);
+
+    MLOG_SET_THREAD_NAME("DL" + std::to_string(thread_id++));
+
+    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&control](){ delete control; });
```

Comment:
And actually I missed that `core::stop` would call cancel, so the instance where the download thread as a pointer to `core` should have been safe. Although it may have been better to additionally call it from `core::~core`, so that the synchronization is not dependent on the code using `core`.